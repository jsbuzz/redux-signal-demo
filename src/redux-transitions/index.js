import { useEffect, useState } from "react";
import { useStore } from "react-redux";

const ACTION_LISTENERS = "_actionListeners";
const IS_THUNK = "_thunkMe";
const thunkRX = /^([^=]+|\([^)]+\))[\s]*=>[\s]*([^=]+|\([^)]+\))[\s]*=>[\s]*|^function[\s]*[^(]*\([^)]*\)[\s]*{[\s]*return[\s]+function[\s]*\([^)]*\)/i;

const thunkKey = (thunkFn) =>
  (`${thunkFn.name}`.length > 3 && thunkFn.name) || thunkFn.toString();

const isThunk = (fn) => fn[IS_THUNK] || !!fn.toString().match(thunkRX);
const asArray = (a) => (Array.isArray(a) ? a : [a]);
const actionKey = (action) =>
  typeof action === "function" ? thunkKey(action) : action.type;
const actionTypeKey = (actionType) =>
  typeof actionType === "function" ? thunkKey(actionType()) : actionType;

// you can use this property in your actions to mark them
// so the middleware will not propagate them to your reducers
export const STOP_PROPAGATION = "_stopPropagation";
export const defaultStates = {
  pending: "pending",
  success: "success",
  failure: "failure",
};

export const createActionListener = () => {
  const context = {};

  return {
    // redux middleware that will call the action listeners
    actionListener: () => (next) => (action) => {
      if (!context.store[ACTION_LISTENERS]) return next(action);

      const key = actionKey(action);
      const listeners = context.store[ACTION_LISTENERS][key];
      if (listeners) {
        listeners.forEach((listener) => listener(action));
      }

      if (!action[STOP_PROPAGATION]) next(action);
    },

    // there is no other way to access the store on the listener level
    setStore: (store) => {
      context.store = store;
    },
  };
};

const processListeners = (listeners, actionCallback) => {
  listeners.reduce((actions, current) => {
    if (typeof current !== "function" || isThunk(current)) {
      return [...actions, current];
    }
    actions.forEach((actionType) =>
      actionCallback(actionTypeKey(actionType), current)
    );
    return [];
  }, []);
};

// hook for setting listeners on actions
export function useActionListeners(...listeners) {
  const store = useStore();
  if (!store[ACTION_LISTENERS]) {
    store[ACTION_LISTENERS] = {};
  }
  useEffect(() => {
    processListeners(listeners, (event, callback) => {
      const actionListeners = store[ACTION_LISTENERS][event];
      if (actionListeners) {
        actionListeners.push(callback);
      } else {
        store[ACTION_LISTENERS][event] = [callback];
      }
    });

    return function cleanup() {
      processListeners(listeners, (event, callback) => {
        const actionListeners = store[ACTION_LISTENERS][event];
        actionListeners.splice(actionListeners.indexOf(callback), 1);
      });
    };
  }, []);
}

// hook for a standard fetch operation with pending and error states
export function usePendingState({
  pending,
  success,
  failure,
  failureHandler = (e) => e,
} = {}) {
  const [state, setState] = useState({
    pending: false,
    error: null,
  });

  useActionListeners(
    ...asArray(pending),
    () =>
      setState({
        pending: true,
        error: null,
      }),

    ...asArray(success),
    () =>
      setState({
        pending: false,
        error: null,
      }),

    ...asArray(failure),
    (errorAction) =>
      setState({
        pending: false,
        error: failureHandler(errorAction),
      })
  );

  return [state.pending, state.error];
}

// hook for transitions
export const useTransitions = (transitionStates, transitionReducer) => {
  const [state, setState] = useState(transitionReducer());

  let listeners = [];
  Object.keys(transitionStates).forEach((transition) => {
    listeners = [
      ...listeners,
      ...asArray(transitionStates[transition]),
      (action) => setState(transitionReducer(transition, action)),
    ];
  });

  useActionListeners(...listeners);

  return state;
};

// function to calculate a transition state from an action
export const mockTransition = (transitionStates, transitionReducer) => (
  action
) => {
  const transition = Object.keys(transitionStates).find(
    (transitionState) =>
      asArray(transitionStates[transitionState]).includes(action) ||
      asArray(transitionStates[transitionState]).includes(action.type)
  );

  return transitionReducer(transition, action);
};

export function thunk(thunkFn) {
  thunkFn[IS_THUNK] = true;
  thunkFn.withTransitionStates = (transitions) => {
    thunkFn.transitions =
      typeof transitions === "function" ? transitions(thunkFn) : transitions;

    return thunkFn;
  };
  return thunkFn;
}

const defaultReducer = (state, error) => [
  state === defaultStates.pending,
  state === defaultStates.failure && error,
];

export function useThunkReducer(thunkFn, reducer) {
  return useTransitions(thunkFn.transitions, reducer || defaultReducer);
}
