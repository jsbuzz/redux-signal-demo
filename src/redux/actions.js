import { STOP_PROPAGATION } from "redux-transitions";

export const FETCH_DATA_START = "FETCH_DATA_START";
export const FETCH_DATA_SUCCESS = "FETCH_DATA_SUCCESS";
export const FETCH_DATA_FAILURE = "FETCH_DATA_FAILURE";

export const fetchDataStart = () => ({
  type: FETCH_DATA_START,
  [STOP_PROPAGATION]: true,
});

export const setData = (data) => ({
  type: FETCH_DATA_SUCCESS,
  data,
});

export const fetchDataError = (error) => ({
  type: FETCH_DATA_FAILURE,
  [STOP_PROPAGATION]: true,
  error,
});

const LOGIN_SUCCESS = "LOGIN_SUCCESS";
const LOGIN_FAILURE = "LOGIN_FAILURE";

export const loginSuccess = (token) => ({
  type: LOGIN_SUCCESS,
  [STOP_PROPAGATION]: true,
  token,
});

export const loginError = (error) => ({
  type: LOGIN_FAILURE,
  [STOP_PROPAGATION]: true,
  error,
});

export const UPLOAD_CHUNK = "UPLOAD_CHUNK";
export const UPLOAD_FAILURE = "UPLOAD_FAILURE";
export const UPLOAD_SUCCESS = "UPLOAD_SUCCESS";

export const addUploadedFile = ({ fileName, fileSize }) => ({
  type: UPLOAD_SUCCESS,
  fileName,
  fileSize,
});

export const uploadError = (error) => ({
  type: UPLOAD_FAILURE,
  [STOP_PROPAGATION]: true,
  error,
});

export const uploadChunk = (percentage) => ({
  type: UPLOAD_CHUNK,
  [STOP_PROPAGATION]: true,
  percentage,
});
