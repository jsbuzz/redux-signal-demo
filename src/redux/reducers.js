import { combineReducers } from "redux";
import { FETCH_DATA_SUCCESS, UPLOAD_SUCCESS } from "./actions";

const dataReducerState = {
  data: null,
};
const dataReducer = (state = dataReducerState, action) => {
  switch (action.type) {
    case FETCH_DATA_SUCCESS:
      return {
        ...state,
        data: action.data,
      };

    default:
      console.log("dataReducer");
      return state;
  }
};

const uploadReducerState = {
  uploadedFiles: [],
  totalSize: 0,
};
const uploadReducer = (state = uploadReducerState, action) => {
  switch (action.type) {
    case UPLOAD_SUCCESS:
      return {
        ...state,
        totalSize: state.totalSize + action.fileSize,
        uploadedFiles: [
          ...state.uploadedFiles,
          {
            fileName: action.fileName,
            fileSize: action.fileSize,
          },
        ],
      };

    default:
      console.log("uploadReducer");
      return state;
  }
};

export default combineReducers({
  data: dataReducer,
  upload: uploadReducer,
});
