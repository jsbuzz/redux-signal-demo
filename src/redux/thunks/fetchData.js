import { fetchDataStart, setData } from "../actions";

export const fetchData = () => (dispatch) => {
  dispatch(fetchDataStart());

  setTimeout(() => {
    dispatch(setData("This is the fetched data"));
  }, 1500);
};
