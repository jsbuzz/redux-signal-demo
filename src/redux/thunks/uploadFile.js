import { addUploadedFile, uploadError, uploadChunk } from "../actions";

let counter = 1;
export const uploadFile = () => (dispatch) => {
  const step = 200 + getRandomInt(400);

  setTimeout(() => dispatch(uploadChunk(15)), step);
  setTimeout(() => dispatch(uploadChunk(35)), 2 * step);
  setTimeout(() => dispatch(uploadChunk(65)), 3 * step);
  setTimeout(() => dispatch(uploadChunk(77)), 4 * step);
  setTimeout(() => dispatch(uploadChunk(85)), 5 * step);
  setTimeout(
    () =>
      dispatch(
        step % 3
          ? addUploadedFile({
              fileName: `${counter++}_somethink.zip`,
              fileSize: (7 * step) / 100,
            })
          : uploadError(`Whoopsie #${counter}`)
      ),
    6.5 * step
  );
};

function getRandomInt(max) {
  return Math.floor(Math.random() * Math.floor(max));
}
