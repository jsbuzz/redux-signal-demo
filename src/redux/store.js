import { createStore, applyMiddleware, compose } from "redux";
import thunk from "redux-thunk";
import { createActionListener } from "../redux-transitions";
import reducers from "./reducers";

const composeEnhancers = window.__REDUX_DEVTOOLS_EXTENSION_COMPOSE__ || compose;

const { actionListener, setStore } = createActionListener();
const store = createStore(
  reducers,
  composeEnhancers(applyMiddleware(actionListener, thunk))
);

setStore(store);

export default store;
