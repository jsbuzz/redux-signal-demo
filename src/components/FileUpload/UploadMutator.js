import React from "react";
import { useDispatch } from "react-redux";
import { useTransitions } from "redux-transitions";
import {
  UPLOAD_CHUNK,
  UPLOAD_SUCCESS,
  UPLOAD_FAILURE,
} from "../../redux/actions";
import { uploadFile } from "../../redux/thunks/uploadFile";

const PENDING_STATE = "pending";
const SUCCESS_STATE = "success";
const FAILURE_STATE = "failure";

const uploadStates = {
  [PENDING_STATE]: [uploadFile, UPLOAD_CHUNK],
  [SUCCESS_STATE]: UPLOAD_SUCCESS,
  [FAILURE_STATE]: UPLOAD_FAILURE,
};

const uploadReducer = (state, { error, percentage }) => ({
  isUploading: state === PENDING_STATE,
  uploadError: state === FAILURE_STATE && error,
  uploadPercentage: percentage || 0,
});

// this component only relies on Transition state
// @transitionState
// @dispatcher
export const UploadMutator = () => {
  const dispatch = useDispatch();

  const { isUploading, uploadPercentage, uploadError } = useTransitions(
    uploadStates,
    uploadReducer
  );

  return (
    <>
      <button disabled={isUploading} onClick={() => dispatch(uploadFile())}>
        {isUploading ? "uploading..." : uploadError ? "retry" : "upload"}
      </button>
      {isUploading && (
        <progress id="file" max="100" value={uploadPercentage}>
          {uploadPercentage}% done
        </progress>
      )}
      {uploadError && (
        <span className="upload-error">
          Error while uploading: {uploadError}
        </span>
      )}
    </>
  );
};
