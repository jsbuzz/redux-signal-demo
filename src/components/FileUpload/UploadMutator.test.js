import React from "react";
import { shallow } from "enzyme";
import { UploadMutator, uploadStates, uploadReducer } from "./UploadMutator";
import * as reduxTransitions from "redux-transitions";
import * as reactRedux from "react-redux";
import { uploadChunk, uploadError } from "../../redux/actions";
import { uploadFile } from "../../redux/thunks/uploadFile";

const mockUploadTransition = reduxTransitions.mockTransition(
  uploadStates,
  uploadReducer
);

describe("UploadMutator", () => {
  let dispatched;
  let wrapper;
  let useTransitionsStub;

  beforeEach(() => {
    dispatched = [];
    jest
      .spyOn(reactRedux, "useDispatch")
      .mockImplementation((action) => dispatched.push(action));
    useTransitionsStub = jest.spyOn(reduxTransitions, "useTransitions");
  });

  describe("Base transition state", () => {
    beforeEach(() => {
      useTransitionsStub.mockReturnValue({});
      wrapper = shallow(<UploadMutator />);
    });

    it("should match snapshot", () => {
      expect(wrapper).toMatchSnapshot();
    });
  });

  describe("Pending transition state start", () => {
    beforeEach(() => {
      useTransitionsStub.mockReturnValue(mockUploadTransition(uploadFile));
      wrapper = shallow(<UploadMutator />);
    });

    it("should match snapshot", () => {
      expect(wrapper).toMatchSnapshot();
    });
  });

  describe("Pending transition state with Chunk", () => {
    const chunk = 37;
    beforeEach(() => {
      useTransitionsStub.mockReturnValue(
        mockUploadTransition(uploadChunk(chunk))
      );
      wrapper = shallow(<UploadMutator />);
    });

    it("should match snapshot", () => {
      expect(wrapper).toMatchSnapshot();
    });
  });

  describe("Error transition state", () => {
    const error = "something went wrong";

    beforeEach(() => {
      useTransitionsStub.mockReturnValue(
        mockUploadTransition(uploadError(error))
      );
      wrapper = shallow(<UploadMutator />);
    });

    it("should match snapshot", () => {
      expect(wrapper).toMatchSnapshot();
    });
  });
});
