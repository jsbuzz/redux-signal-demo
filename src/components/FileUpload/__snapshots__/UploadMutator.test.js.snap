// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`UploadMutator Base transition state should match snapshot 1`] = `
<Fragment>
  <button
    onClick={[Function]}
  >
    upload
  </button>
</Fragment>
`;

exports[`UploadMutator Error transition state should match snapshot 1`] = `
<Fragment>
  <button
    disabled={false}
    onClick={[Function]}
  >
    retry
  </button>
  <span
    className="upload-error"
  >
    Error while uploading: 
    something went wrong
  </span>
</Fragment>
`;

exports[`UploadMutator Pending transition state with \`uploadFile\` should match snapshot 1`] = `
<Fragment>
  <button
    disabled={true}
    onClick={[Function]}
  >
    uploading...
  </button>
  <progress
    id="file"
    max="100"
    value={0}
  >
    0
    % done
  </progress>
</Fragment>
`;

exports[`UploadMutator Pending transition state with Chunk should match snapshot 1`] = `
<Fragment>
  <button
    disabled={true}
    onClick={[Function]}
  >
    uploading...
  </button>
  <progress
    id="file"
    max="100"
    value={37}
  >
    37
    % done
  </progress>
</Fragment>
`;
